Decode custom name section

ColinEberhardt opened the pull request:
This PR adds logic to the wasm-parser to decode custom name sections, fixes #202 

Changes in this PR:
 - wasm-parser tests no longer have binary wasm files as test cases, they instead use wat files which are transformed to wasm via `wat2wasm` (I'm not sure how we make this a proper dependency?)
 - wasm-parser parses the custom name section generated by `wat2wasm --debug-names`. Currently it only parses the function name sub-section. I'll look at adding local names next. Names are stored in the module-metadata (which I've restructured a bit to make it easier to walk).
 - wasm-parser walks the AST, after decoding, to apply names. I've added an option to disable this behaviour as desired, as there will be a performance impact.

TODO:
 - [x] work out how make `wat2wasm` available to travis
 - [x] fix flow errors!

xtuc commented on packages/ast/src/index.js:
```diff
@@ -33,6 +33,32 @@ export function signature(object: string, name: string): SignatureMap {
+    n.functionNames = functionNames;
+  }
+  return n;
+}
```
I'm not sure the metadata should be a node in our AST? I already model it here https://github.com/xtuc/webassemblyjs/blob/7b3e82af4bf89b49af46cac5d64e8659a06b8037/packages/ast/src/index.js#L93 as a property on the module, what do you think?

ColinEberhardt replied:
I've kept the metadata in the same place. The change I've made is to give it a `type` property so that it can be traversed by the AST parser. Does that sound OK?

xtuc commented on packages/wasm-parser/src/decoder.js:
```diff
@@ -1388,17 +1447,23 @@ export function decode(ab: ArrayBuffer, opts: DecoderOpts): Program {
+        if (sectionName.value === "name") {
+          metadata.push(...parseNameSection(remainingBytes));
+        } else {
+          // We don't parse otehr custom section
```
nit: typo `otehr` -> `other`

xtuc commented on packages/wasm-parser/src/decoder.js:
```diff
@@ -959,6 +962,59 @@ export function decode(ab: ArrayBuffer, opts: DecoderOpts): Program {
+
+  // this is a custom suction that wat2wasm includes if invoked
+  // using the --debug-names option
+  function parseNameSection(remainingBytes: number) {
```
I'm wondering if we shouldn't split it in a `wabt` file or something? Because it's specific to the binary they're emitting.

Also how we could know that it has been emitted by `wabt`?

ColinEberhardt replied:
good point - my next step will be to complete this work by supporting locals, so I could have a go at that then?

xtuc replied:
Yes sure

xtuc commented on packages/wasm-parser/src/index.js:
```diff
@@ -1,14 +1,51 @@
+  ignoreCustomNameSection: false
 };
 
+function restoreNames(ast) {
```
Instead of doing a separate pass couldn't we do this in the decoder directly?

ColinEberhardt replied:
I don't think this is possible. The name section appears at the very end of the module, so the only way to apply the naming is via a second pass.

xtuc replied:
Ok that make sense.

According to the spec a custom section can be after each non-custom section, so depending on how the binary was constructed It could happens, but we can leave as is since it will work in any case.

xtuc commented:
>  work out how make wat2wasm available to travis

We could use https://www.npmjs.com/package/wast2wasm and wrap it into an executable.

ColinEberhardt commented:
@xtuc I've updated and fixed the tests - it now uses wabt.js rather than the executable. There are a few unresolved discussion points, would be interested in your feedback.

After this, I'll look at decoding local names.

Then, the final goal, run the interpreter test fixtures,  against the wasm-parser!

xtuc commented on packages/wasm-parser/src/decoder.js:
```diff
@@ -959,6 +962,59 @@ export function decode(ab: ArrayBuffer, opts: DecoderOpts): Program {
+    const numbeOfFunctions = numbeOfFunctionsu32.value;
+    eatBytes(numbeOfFunctionsu32.nextIndex);
+
+    for (let i = 0; i < numbeOfFunctions; i++) {
```
Here's a small typo: `numbeOfFunctions` -> `numberOfFunctions`

ColinEberhardt commented:
Ooops - I made a mistake in the files which I was round-tripping. Need to make a few more changes. Although now that the tests are correct it is starting to throw up differences in the wat and wasm parsers:

https://github.com/xtuc/webassemblyjs/issues/218

ColinEberhardt commented:
I've simplified this PR to just include the custom name section logic - I'll add the wasm-parser test work to a different PR as this is starting to look a like a bigger job
